source/common/config/delta_subscription_impl.h

```diff
@@ -105,38 +96,37 @@ class DeltaSubscriptionImpl : public Subscription,
     ENVOY_LOG(debug, "Resuming discovery requests for {}", type_url_);
     ASSERT(paused_);
     paused_ = false;
-    if (pending_.has_value()) {
-      queueDiscoveryRequest(pending_.value());
-      pending_.reset();
-    }
+    trySendDiscoveryRequests();
   }
 
   envoy::api::v2::DeltaDiscoveryRequest internalRequestStateForTest() const { return request_; }
 
   // Config::GrpcStreamCallbacks
   void onStreamEstablished() override {
-    // initial_resource_versions "must be populated for first request in a stream", so guarantee
-    // that the initial version'd request we're about to enqueue is what gets sent.
-    clearRequestQueue();
-
     request_.Clear();
     for (auto const& resource : resource_versions_) {
       // Populate initial_resource_versions with the resource versions we currently have. Resources
       // we are interested in, but are still waiting to get any version of from the server, do not
-      // belong in initial_resource_versions.
+      // belong in initial_resource_versions. (But do belong in new subscriptions!)
       if (!resource.second.waitingForServer()) {
         (*request_.mutable_initial_resource_versions())[resource.first] = resource.second.version();
       }
+      // As mentioned above, fill resource_names_subscribe with everything.
+      names_added_.insert(resource.first);
     }
+    names_removed_.clear();
     request_.set_type_url(type_url_);
     request_.mutable_node()->MergeFrom(local_info_.node());
-    queueDiscoveryRequest(ResourceNameDiff()); // no change to subscribed resources
+    kickOffDiscoveryRequest();
   }
 
   void onEstablishmentFailure() override {
     disableInitFetchTimeoutTimer();
     stats_.update_failure_.inc();
     ENVOY_LOG(debug, "delta update for {} failed", type_url_);
+    // TODO(fredlas) this increment is needed to pass existing tests, but it seems wrong. We already
+    // increment it when updating subscription interest, which attempts a request. Is this supposed
+    // to be the sum of client- and server- initiated update attempts? Seems weird.
```

Comment:
I agree, that seems like a bug.

Reply:
@eziskind actually independently ran into this in the existing code and thought it was weird, the other day, and we talked about it a bit.

Reply:
@eziskind and I were also discussing this recently